pkg/store/postings_codec.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"bytes"
+
+	"github.com/golang/snappy"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/encoding"
+	"github.com/prometheus/prometheus/tsdb/index"
+)
+
+const (
+	// These headers should not be a prefix of each other.
+	codecHeaderRaw    = "diff+varint+raw"
+	codecHeaderSnappy = "diff+varint+snappy"
+)
+
+// isDiffVarintEncodedPostings returns true, if input looks like it has been encoded by diff+varint(+snappy) codec.
+func isDiffVarintEncodedPostings(input []byte) bool {
+	return bytes.HasPrefix(input, []byte(codecHeaderRaw)) || bytes.HasPrefix(input, []byte(codecHeaderSnappy))
+}
+
+// diffVarintEncode encodes postings into diff+varint representation and optional snappy compression.
+func diffVarintEncode(p index.Postings, useSnappy bool) ([]byte, error) {
+	buf := encoding.Encbuf{}
+
+	// If we're returning raw data, write the header to the buffer, and then return buffer directly.
+	if !useSnappy {
+		buf.PutString(codecHeaderRaw)
+	}
+
+	prev := uint64(0)
+	for p.Next() {
+		v := p.At()
+		if v < prev {
+			// Postings entries must be in increasing order.
+			return nil, errors.Errorf("decreasing entry, val: %d, prev: %d", v, prev)
+		}
+
+		buf.PutUvarint64(v - prev)
```

Comment:
why we do delta? For better compression?

Reply:
This one is very much needed to be commented then ;p

Reply:
Snappy compression on its own doesn't really help, because there is no repetition. Delta makes values small, and Varint is very efficient at encoding small values (values < 128 are encoded as single byte, values < 16384 are encoded as two bytes). It's delta + using varint encoding that reduces size significantly (to about 20% of original), snappy then halves it to 10% of the original.

Reply:
Well, "diff" (or "delta", perhaps we should rename it) is already a strong indication of what we're doing here :)

Reply:
:+1: Thanks!

Reply:
Maybe strong indication but what you said here:

> Snappy compression on its own doesn't really help, because there is no repetition. Delta makes values small, and Varint is very efficient at encoding small values (values < 128 are encoded as single byte, values < 16384 are encoded as two bytes). It's delta + using varint encoding that reduces size significantly (to about 20% of original), snappy then halves it to 10% of the original.

Is important and will be forgotten if it's only on GH PR comment.

Reply:
Fair enough, I'll put this comment into the file.